cmd/medius/images/promote.go

```diff
@@ -0,0 +1,103 @@
+package images
+
+import (
+	"context"
+	"errors"
+	"path"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"kubevirt.io/containerdisks/cmd/medius/common"
+	"kubevirt.io/containerdisks/pkg/api"
+	"kubevirt.io/containerdisks/pkg/repository"
+)
+
+func NewPromoteImagesCommand(options *common.Options) *cobra.Command {
+	options.PromoteImageOptions = common.PromoteImageOptions{
+		TargetRegistry: "quay.io/containerdisks",
+	}
+
+	promoteCmd := &cobra.Command{
+		Use:   "promote",
+		Short: "Promote verified containerdisks from one registry to another registry",
+		Run: func(cmd *cobra.Command, args []string) {
+			results, err := readResultsFile(options.ImagesOptions.ResultsFile)
+			if err != nil {
+				logrus.Fatal(err)
+			}
+
+			resultsChan, err := spawnWorkers(cmd.Context(), options, func(a api.Artifact) (*api.ArtifactResult, error) {
+				r, ok := results[a.Metadata().Describe()]
+				if !ok || r.Stage != StageVerify || r.Err != "" {
+					return nil, nil
+				}
+
+				errString := ""
+				err := promoteArtifact(cmd.Context(), a, r.Tags, options)
+				if err != nil {
+					errString = err.Error()
+				}
+
+				return &api.ArtifactResult{
+					Tags:  r.Tags,
+					Stage: StagePromote,
+					Err:   errString,
+				}, err
+			})
+
+			for result := range resultsChan {
+				results[result.Key] = result.Value
+			}
+
+			if !options.DryRun {
+				if err := writeResultsFile(options.ImagesOptions.ResultsFile, results); err != nil {
+					logrus.Fatal(err)
+				}
+			}
+
+			if err != nil {
+				logrus.Fatal(err)
+			}
+		},
+	}
+	promoteCmd.Flags().StringVar(&options.PromoteImageOptions.SourceRegistry, "source-registry", options.PromoteImageOptions.SourceRegistry, "Registry to pull images from")
+	promoteCmd.Flags().StringVar(&options.PromoteImageOptions.TargetRegistry, "target-registry", options.PromoteImageOptions.TargetRegistry, "Registry to promote images to")
+
+	err := promoteCmd.MarkFlagRequired("source-registry")
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	return promoteCmd
+}
+
+func promoteArtifact(ctx context.Context, artifact api.Artifact, tags []string, options *common.Options) error {
+	log := common.Logger(artifact)
+
+	if len(tags) == 0 {
+		err := errors.New("No containerdisks to promote")
```

Comment:
If all images are in sync, wouldn't it be the expected case that we don't have any work to do? If this is about having something to promote but no tags, we probably have a bug and want a different error?

Reply:
If all images are in sync `images push` should omit the artifact from the results file. But you are right, a check in the code for that is missing.